compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -3616,8 +3638,24 @@ private boolean isAssignableToUnionType(BType source, BType target, Set<TypePair
                     break;
                 }
 
-                if (isAssignable(sourceMember, targetMember, unresolvedTypes)) {
-                    sourceTypeIsNotAssignableToAnyTargetType = false;
+                if (sourceMember.tag == TypeTags.ERROR) {
+                    if (targetUnionIsDistinctError) {
+                        sourceTypeIsNotAssignableToAnyTargetType = !isErrorTypeAssignable((BErrorType) sourceMember,
+                                (BErrorType) targetMember, unresolvedTypes);
+                    } else if (allTargetTypesErrors && targetMember.tag == TypeTags.ERROR) {
+                        sourceTypeIsNotAssignableToAnyTargetType =
+                                !isErrorTypeAssignableWithNoTypeIds((BErrorType) sourceMember,
+                                        (BErrorType) targetMember, unresolvedTypes);
+                    } else {
+                        sourceTypeIsNotAssignableToAnyTargetType = !isAssignable(sourceMember, targetMember,
+                                unresolvedTypes);
+                    }
+                } else {
+                    sourceTypeIsNotAssignableToAnyTargetType = !isAssignable(sourceMember, targetMember,
+                            unresolvedTypes);
+                }
+
+                if (!sourceTypeIsNotAssignableToAnyTargetType) {
```

Comment:
We are inside a while loop for `targetIterator`. Is it correct to break here if this variable is false? If that is the case IMO it is better to rename the variable. If we break here it is because a source type is not assignable to one of the target types right?

Variable is intended for line 3662 which is after we check a particular source against all the targets.